### ISSUE TITLE
MultiProcessingReadingService: High communication overhead

### ISSUE THREAD

**Issue by sehoffmann** [pytorch/data#1078]
### 🐛 Describe the bug

I've recently came to test and benchmark torchdata and stumbled upon a significant overhead when working with MultiProcessingReadingService. Unfortunately, my pipeline contains some custom operations acting on xarray Datasets; but nothing fancy is being done. Essentially just some preloading, indexing, and cropping. I will post it regardless for completeness sake:

```
years = [1990, 1991]
variables = ['z', 't', 'r']
N_vars = len(variables)

def open_variable(var, pipe):
    pipe = pipe.wb_build_path(base_dir, var).xr_open()
    return pipe.xr_get_variables(var).xr_sel(level=500).xr_split_dim(4, 'time')

pipes = dp.iter.IterableWrapper(years).fork(N_vars,5)
pipe = dp.iter.Zipper(*[open_variable(var,pipe) for var, pipe in zip(variables, pipes)])
pipe = pipe.sharding_filter()
pipe = pipe.xr_prefetch(executor=ThreadPoolExecutor)
pipe = pipe.xr_merge()
pipe = pipe.xr_extract_timeseries(10, 6)
pipe = pipe.repeat(4)
pipe = pipe.xr_random_crop({'lat': 96, 'lon': 96}, wraps={'lon': True})
pipe = pipe.shuffle(buffer_size=10000).batch(32)

from torchdata.dataloader2 import MultiProcessingReadingService, DataLoader2

rs = MultiProcessingReadingService(num_workers=1)
dl = DataLoader2(pipe, reading_service=rs)
```

I've compared the timings per batch for not using `DataLoader2`, and with n number of workers respectively. When using multiple IO workers, I divide the shuffle buffer size by it.

##### Only Pipe
* 258 s total
* 121 ms / batch (32)
* 48 s latency

##### 1 Worker
* 467s total
* 219 ms / batch (32)
* 49 s latency

##### 2 Worker
* 275s total 
* 129 ms / batch (32)
* 27 s latency

##### 4 Worker
* 224s total
* 105 ms / batch (32)
* 23 s latency

As you can see, there is a very significant performance overhead of 81% if one uses a single IO worker oppose to none at all. This destroys many of the gains that could be achieved from multiprocessing and only with 4 workers I see a small speedup of  15% which mostly seem to be due to the reduction of latency (time for shuffle buffer fill and initial preloading) though.

### Questions:
1. I've not yet converted the `xarray.DataArray`s to `torch.Tensor`s; could this be a cause for such severe overheads when transmitting via IPC?
2. Should I decide to use `torch.Tensor`, can both the main and worker processes communicate via shared memory such that only the meta information need to be transmitted (i.e. essentially a pointer + shape + stridings  etc.)
3. Can `torch.Tensor`s preloaded on gpu device memory be transmitted such that only a cuda pointer needs to be transmitted?

### Versions

```
Collecting environment information...
PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (conda-forge gcc 11.3.0-19) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.17

Python version: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.76.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                72
On-line CPU(s) list:   0-71
Thread(s) per core:    2
Core(s) per socket:    18
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz
Stepping:              7
CPU MHz:               3234.448
CPU max MHz:           3900.0000
CPU min MHz:           1000.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-17,36-53
NUMA node1 CPU(s):     18-35,54-71
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin intel_pt ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] numpy-groupies==0.9.20
[pip3] torch==1.13.1
[pip3] torchdata==0.5.1
[pip3] torchvision==0.14.1
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] cudatoolkit               11.7.0              hd8887f6_10    nvidia
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.23.5          py310h53a5b5f_0    conda-forge
[conda] numpy_groupies            0.9.20             pyhd8ed1ab_0    conda-forge
[conda] pytorch                   1.13.1          py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h67b0de4_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchdata                 0.5.1                    pypi_0    pypi
[conda] torchvision               0.14.1              py310_cu117    pytorch

```

**Comment by ejguan** [comment]
I am not sure how complicated of those operations
```py
pipe = pipe.xr_prefetch(executor=ThreadPoolExecutor)
pipe = pipe.xr_merge()
pipe = pipe.xr_extract_timeseries(10, 6)
pipe = pipe.repeat(4)
pipe = pipe.xr_random_crop({'lat': 96, 'lon': 96}, wraps={'lon': True})
```
If they are CPU heavy, MP should give better perf. Could you please explain a little bit on the `latency`?

And, in terms of inter-process communication, it only happens at the end of your pipeline. So, it's better to convert the `batch` to a `Tensor`-like object to pass IMHO.

And, to validate your benchmarking result, could you pls add a timer around the first few warm-up iterations and a time after the warm-up iterations. When MP gets involved, we automatically add `prefetch`, which takes some time to build up the buffer.

**Comment by sehoffmann** [comment]
Hey @ejguan,

So first up: I'm very confident that my pipeline is CPU-bound as I not only have a high data reuse factor but also because if I monitor the IO throughput (I'm reading from network storage by the way) of my processes with htop, I see very short spikes with almost max IO throughput (500-600mb/s) and then very long periods with zero IO activity. This expected and perfectly coincidents with the prefetching behavior.

Wrt. latency: This is time required for the first batch to arrive. This time is mostly comprised of the time required to fill up the shuffle buffer, but also prefetch buffers, and also the time until we've loaded the first files into memory (happens in `xr_prefetch`). I'm actually storing and plotting the time for each individual batch, and as such can confidently say that after this initial delay for the first batch, the time required for subsequent batches is closely distributed around the mean.

----------------------------------------

I made some small changes to the code above:
1. I got rid of the shuffle buffer in order to get it out of the equation for benchmarking purposes
2. I now proper collate my `xarray.Dataset`s to `torch.Tensor`s. 

For completeness sake, this is the new pipe:
```
base_dir = '/mnt/qb/datasets/WeatherBench/1.40625deg'
years = [1990]
variables = ['z', 't', 'r']
N_vars = len(variables)

def open_variable(var, pipe):
    pipe = pipe.wb_build_path(base_dir, var).xr_open()
    return pipe.xr_get_variables(var).xr_sel(level=500).xr_split_dim('time', 12)

pipes = dp.iter.IterableWrapper(years).fork(N_vars,5)
pipe = dp.iter.Zipper(*[open_variable(var,pipe) for var, pipe in zip(variables, pipes)])
pipe = pipe.sharding_filter()
pipe = pipe.xr_prefetch(executor=ThreadPoolExecutor)
pipe = pipe.xr_merge()
pipe = pipe.xr_extract_timeseries(10, 6)
pipe = pipe.repeat(4)
pipe = pipe.xr_random_crop({'lat': 96, 'lon': 96}, wraps={'lon': True})
pipe = pipe.xr_to_numpy().batch(32).collate()

from torchdata.dataloader2 import MultiProcessingReadingService, DataLoader2

rs = MultiProcessingReadingService(num_workers=4)
dl = DataLoader2(pipe, reading_service=rs)
```

Here are the new results:
##### Pipe Only
* 289s total
* 287 ms / batch (32)
* 2.5s latency

##### 1 Worker
* 281s total
* 279 ms / batch (32)
* 3.4s latency

##### 2 Worker
* 158s total
* 157 ms / batch (32)
* 2.9s latency

##### 4 Workers

* 93s total
* 93 ms / batch (32)
* 2.5s latency


-----------------------------------

We can see two major things happening:
1. The average throughput for the "Only Pipe" and "1 Worker" case is drastically reduced.
2. However, MPRS now properly scales and there is almost now performance difference between iterating the pipe directly and using only 1 worker.

What I suspect happened before is that `xr_extract_timeseries()` and `xr_random_crop()` mostly produced views into already preloaded data arrays which then got placed into the shuffle buffer. Once MPRS got turned on, these views now actually got copied/accessed since they had to be serialized. As a result, performance degraded drastically. Now with the collate function explicitly in place, these copies also happen for the non-MPRS case, which also explains why performance degraded in general.

One thing to note here, is that `default_collate` seem to be quite slow for single elements and its much better to invoke it on lists/batches. Maybe this should be noted in the documentation. I didn't perform extensive experimentation on this yet though.

**Comment by ejguan** [comment]
I personally think the perf is affected by python objects are slow to be passes across processes but it requires copy-paste. But, Tensor would be shared via IPC or file handles.

For single element, you might want to just do a `torch.tensor(x)` in a `dp.map()` to accelerate the pipeline.

**Comment by sehoffmann** [comment]
@ejguan Yeah you are absolutely right. Transmitting numpy.arrays instead hurt performance significantly (50% of original throughput). I believe it would be good to mention these performance considerations in the documentation for the MPRS since they really have very significant implications for throughput.
